src/ConstrainedPhotometryModel.cc

```diff
@@ -0,0 +1,143 @@
+            _chipMap[chip] = std::unique_ptr<PhotometryMapping>(new PhotometryMapping(chipTransfo));
+        }
+        // If the visit is not in the map, add it, otherwise continue.
+        if (visitPair == _visitMap.end()) {
+            auto visitTransfo = std::make_shared<PhotometryTransfoChebyshev>(visitDegree, focalPlaneBBox);
+            _visitMap[visit] = std::unique_ptr<PhotometryMapping>(new PhotometryMapping(visitTransfo));
+        }
+    }
+
+    // Fix one chip mapping, to remove the degeneracy from the system.
+    _chipMap.at(constrainedChip)->setFixed(true);
+
+    // Now create the ccdImage mappings, which are combinations of the chip/visit mappings above.
+    _myMap.reserve(ccdImageList.size());  // we know how big it will be, so pre-allocate space.
+    for (auto const &ccdImage : ccdImageList) {
+        auto visit = ccdImage->getVisit();
+        auto chip = ccdImage->getCcdId();
+        _myMap.emplace(ccdImage->getHashKey(),
+                       std::unique_ptr<ChipVisitPhotometryMapping>(
+                               new ChipVisitPhotometryMapping(_chipMap[chip], _visitMap[visit])));
+    }
+    LOGLS_INFO(_log, "Got " << _chipMap.size() << " chip mappings and " << _visitMap.size()
+                            << " visit mappings; holding chip " << constrainedChip << " fixed.");
+    LOGLS_DEBUG(_log, "CcdImage map has " << _myMap.size() << " mappings, with " << _myMap.bucket_count()
+                                          << " buckets and a load factor of " << _myMap.load_factor());
+}
+
+unsigned ConstrainedPhotometryModel::assignIndices(std::string const &whatToFit, unsigned firstIndex) {
+    // TODO: currently ignoring whatToFit: eventually implement configurability.
+    unsigned index = firstIndex;
+    for (auto &i : _chipMap) {
+        auto mapping = i.second.get();
+        // Don't assign indices for fixed parameters.
+        if (mapping->isFixed()) continue;
+        mapping->setIndex(index);
+        index += mapping->getNpar();
+    }
+    for (auto &i : _visitMap) {
+        auto mapping = i.second.get();
+        mapping->setIndex(index);
+        index += mapping->getNpar();
+    }
+    return index;
+}
+
+void ConstrainedPhotometryModel::offsetParams(Eigen::VectorXd const &delta) {
+    for (auto &i : _chipMap) {
+        auto mapping = i.second.get();
+        // Don't offset indices for fixed parameters.
+        if (mapping->isFixed()) continue;
+        mapping->offsetParams(delta.segment(mapping->getIndex(), mapping->getNpar()));
+    }
+    for (auto &i : _visitMap) {
+        auto mapping = i.second.get();
+        mapping->offsetParams(delta.segment(mapping->getIndex(), mapping->getNpar()));
+    }
+}
+
+double ConstrainedPhotometryModel::transform(CcdImage const &ccdImage, MeasuredStar const &measuredStar,
+                                             double instFlux) const {
+    auto mapping = this->findMapping(ccdImage, "transform");
```

Comment:
I removed all `this->` as part of one commit, but then had to add a few of them back in because the compiler was unhappy.

I didn't see any guidelines on what to use `this->` in the dev guide: is it basically "only when the code won't compile without it"?